xarray/core/indexing.py

```diff
@@ -275,25 +276,40 @@ def remap_label_indexers(data_obj, indexers, method=None, tolerance=None):
     return pos_indexers, new_indexes
 
 
+def _normalize_slice(sl, size):
+    """Ensure that given slice only contains positive start and stop values
+    (stop can be -1 for full-size slices with negative steps, e.g. [-10::-1])"""
+    return slice(*sl.indices(size))
+
+
 def slice_slice(old_slice, applied_slice, size):
     """Given a slice and the size of the dimension to which it will be applied,
     index it with another slice to return a new slice equivalent to applying
     the slices sequentially
     """
-    step = (old_slice.step or 1) * (applied_slice.step or 1)
-
-    # For now, use the hack of turning old_slice into an ndarray to reconstruct
-    # the slice start and stop. This is not entirely ideal, but it is still
-    # definitely better than leaving the indexer as an array.
-    items = _expand_slice(old_slice, size)[applied_slice]
-    if len(items) > 0:
-        start = items[0]
-        stop = items[-1] + int(np.sign(step))
-        if stop < 0:
-            stop = None
-    else:
-        start = 0
-        stop = 0
+    old_slice = _normalize_slice(old_slice, size)
+
+    size_after_old_slice = math.ceil(
+        (old_slice.stop - old_slice.start) / old_slice.step
+    )
```

Comment:
I would have used `len(range(old_slice.start, old_slice.stop, old_slice.step))` here but timing shows that we'd lose about 30ns (on my machine), which I assume is the time it takes to construct the `range` object.

Reply:
Oh but that's so much less error-prone! I'd say kill the nanoseconds.